src/groupchat_packets.md

```diff
@@ -0,0 +1,1128 @@
+- if `event` is MV_BAN then
+    - parse `sanction list`.
+    - add sanction list.
+- call callback of client to update UI.
+- delete peer from group.
+
+##### Remove ban
+
+Serialized form:
+
+Length      | Content
+------------| ------
+`4`         | `ban_id`
+variable    | `sanction list`(see above)
+
+When a peer receives this packet then:
+
+- parse sanction list.
+- remove sanctions which is listed in sanction list.
+
+##### Set/unset moderator
+
+Serialized form:
+
+Length      | Content
+------------| ------
+`1`         | `kind`(0 = set to user, otherwise = set to moderator)
+`32`        | PK(`kind` = 0) or Moderator hash(`kind` is not 0)
+
+When a peer receives this packet then:
+
+- if `kind` is not `0` then
+    - get peer number using moderator hash.
+    - add peer to moderator list.
+    - set peer's role to moderator.
+- else
+    - get peer number using PK.
+    - remove peer from moderator list.
+    - set peer's role to user.
+- call callback of client to update UI.
+
+##### Set observer
+
+Serialized form:
+
+Length      | Content
+------------| ------
+`1`         | `kind`(0 = remove sanctions, otherwise = set sanctions)
+`32`        | PK
+variable    | `sanction list`(see above)
+
+When a peer receives this packet then:
+
+- if `kind` is not 0 then
+    - add sanction list to group.
+-else
+    - remove sanction list from group.
+- call callback of client to update UI.
+
+#### Peer announce (0xf2)
+
+Serialized form:
+
+Length      | Content
+------------|-------
+`32`        | PK of peer
+`1`         | `flag`(of ip port is setted)
+`1`         | `count`(of tcp relays)
+`1`         | `type`(of ip)
+`4` or `16` | IPv4 or IPv6 address(comes only when `flag` is setted)
+`2`         | port(comes only when `flag` is setted)
+variable    | tcp relay list
+
+An entry of `tcp relay list` is
+
+Length      | Content
+------------|-------
+`1`         | `type`(of ip)
+`4` or `16` | IPv4 or IPv6 address
+`2`         | port
+`32`        | PK of peer
+
+When a peer receives this packet then:
+
+- if the peer is in sanction list then
+    - return error.
+- add peer to group with PK and Ip, Port.
+- add and save tcp relays to group data structures.
+
+#### Peer info request (0xf4)
+
+This packet has no data body.
+
+When a peer receives this packet then:
+
+- send self info to peer.
+
+#### Peer info response (0xf5)
+
+Serialized form:
+
+Length      | Content
+------------|-------
+`32`        | `password`(comes only when password is setted)
+`2`         | `length` of nickname
+`128`       | `nickname`(UTF-8 string)
+`1`         | `status`
+`1`         | `role`
+
+When a peer receives this packet then:
+
+- updatee info of the peer.
+- validate role of peer, if it is error then
+    - return error.
+- call callback of client to update UI.
+
+#### Sync Request (0xf8)
+
+Serialized form:
+
+Length      | Content
+------------|-------
+`32`        | `password`(comes only when password is setted)
```

Comment:
> password

plaintext?

Reply:
Yes.
It was encrypted with shared secret and there is no additional encryption.